mos-platform/rp6502/rp6502.h

```diff
@@ -0,0 +1,197 @@
+#ifndef _RP6502_H
+#define _RP6502_H
+
+/* RP6502 RIA $FFE0-$FFF9 */
+struct __RP6502 {
+  const unsigned char ready;
+  unsigned char tx;
+  const unsigned char rx;
+  const unsigned char vsync;
+  unsigned char rw0;
+  unsigned char step0;
+  unsigned int addr0;
+  unsigned char rw1;
+  unsigned char step1;
+  unsigned int addr1;
+  unsigned char xstack;
+  unsigned char errno_lo;
+  unsigned char errno_hi;
+  unsigned char op;
+  unsigned char irq;
+  const unsigned char spin;
+  const unsigned char busy;
+  const unsigned char lda;
+  unsigned char a;
+  const unsigned char ldx;
+  unsigned char x;
+  const unsigned char rts;
+  unsigned int sreg;
+};
+#define RIA (*(volatile struct __RP6502 *)0xFFE0)
+
+#define RIA_READY_TX_BIT 0x80
+#define RIA_READY_RX_BIT 0x40
+#define RIA_BUSY_BIT 0x80
+
+/* OS operation numbers */
+
+#define RIA_OP_EXIT 0xFF
+#define RIA_OP_ZXSTACK 0x00
+#define RIA_OP_XREG 0x01
+#define RIA_OP_PHI2 0x02
+#define RIA_OP_CODEPAGE 0x03
+#define RIA_OP_LRAND 0x04
+#define RIA_OP_STDIN_OPT 0x05
+#define RIA_OP_CLOCK 0x0F
+#define RIA_OP_CLOCK_GETRES 0x10
+#define RIA_OP_CLOCK_GETTIME 0x11
+#define RIA_OP_CLOCK_SETTIME 0x12
+#define RIA_OP_CLOCK_GETTIMEZONE 0x13
+#define RIA_OP_OPEN 0x14
+#define RIA_OP_CLOSE 0x15
+#define RIA_OP_READ_XSTACK 0x16
+#define RIA_OP_READ_XRAM 0x17
+#define RIA_OP_WRITE_XSTACK 0x18
+#define RIA_OP_WRITE_XRAM 0x19
+#define RIA_OP_LSEEK 0x1A
+#define RIA_OP_UNLINK 0x1B
+#define RIA_OP_RENAME 0x1C
+
+/* C API for the operating system. */
+
+int xregn(char device, char channel, unsigned char address, unsigned char count,
+          ...);
+int phi2 (void);
+int codepage (void);
+long lrand (void);
+int stdin_opt (unsigned long ctrl_bits, unsigned char str_length);
+
+/* XREG location helpers */
+
+#define xreg_ria_keyboard(...) xreg(0, 0, 0, __VA_ARGS__)
+#define xreg_ria_mouse(...) xreg(0, 0, 1, __VA_ARGS__)
+#define xreg_vga_canvas(...) xreg(1, 0, 0, __VA_ARGS__)
+#define xreg_vga_mode(...) xreg(1, 0, 1, __VA_ARGS__)
+
+/* XRAM structure helpers */
+
+#define xram0_struct_set(addr, type, member, val)                              \
```

Comment:
This macro and the following should really be functions. If it's desired that they always be inlined, then it's sufficient to mark them with `__attribute__((always_inline))`.

Reply:
Because `type` is a struct name and `member` is a struct member name, the best I can do is split this into macros *and* inlines. That makes the code harder to follow with no benefit to the compiled output.

Reply:
Sorry, missed type; my bad.